src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+            // we successfully send or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            var writtenBuffer = _writeBuffer;
+            _writeBuffer = null;
+
+            using (cancellationToken.Register(() => tcsSend.TrySetCanceled()))
+            {
+                try
+                {
+                    if (messageType == WebSocketMessageType.Binary)
+                    {
+                        using (var uint8Buffer = Uint8Array.From(buffer))
+                        {
+                            _innerWebSocket!.Invoke("send", uint8Buffer);
+                            tcsSend.SetResult(true);
+                        }
+                    }
+                    else if (messageType == WebSocketMessageType.Text)
+                    {
+                        string strBuffer = buffer.Array == null ? string.Empty : Encoding.UTF8.GetString(buffer.Array, buffer.Offset, buffer.Count);
+                        _innerWebSocket!.Invoke("send", strBuffer);
+                        tcsSend.SetResult(true);
+                    }
+                }
+                catch (Exception excb)
+                {
+                    tcsSend.TrySetException(new WebSocketException(WebSocketError.NativeError, excb));
+                }
+                finally
+                {
+                    writtenBuffer?.Dispose();
+                }
+                await tcsSend.Task.ConfigureAwait(continueOnCapturedContext: true);
+            }
+        }
+
+        /// <summary>
+        /// Receives data on <see cref="System.Net.WebSockets.ClientWebSocket"/> as an asynchronous operation.
+        /// </summary>
+        /// <returns>The async.</returns>
+        /// <param name="buffer">Buffer.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        public override async Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            ThrowOnInvalidState(State, WebSocketState.Open, WebSocketState.CloseSent);
+
+            var tcsReceive = new TaskCompletionSource<WebSocketReceiveResult>();
+
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully receive or not.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            using (cancellationToken.Register(() => tcsReceive.TrySetCanceled()))
+            {
+                if (_bufferedPayload == null)
+                    _bufferedPayload = await _receiveMessageQueue.DequeuePayloadAsync(cancellationToken).ConfigureAwait(continueOnCapturedContext: true);
+
+                try
+                {
+                    var endOfMessage = _bufferedPayload.BufferPayload(buffer, out WebSocketReceiveResult receiveResult);
+
+                    tcsReceive.SetResult(receiveResult);
```

Comment:
This is going to throw an exception if cancellation occurs and completes the task first.

Reply:
Stephen

The TCS is needed because we do not have threading enabled at this time.  The only way to have await work is to use TaskCompletionSource to control when the task should return with the result.

Reply:
That explanation doesn't make sense to me.  You have the receiveResult here.  Why can't you just `return receiveResult;`?

Reply:
yes I see what you mean now.

Reply:
Addressed in recent commit